### PR TITLE
State handling updates for Reader/Writer Groups and DataSetReader/Writer

### PR DESCRIPTION
**Pull request by gauravks**
- Add support for Pre-operational states
- Add API to enable the objects

Replaces #5456

### CODE REVIEW

File: include/open62541/server_pubsub.h [hunk 1]
```diff
@@ -573,6 +575,9 @@ UA_EXPORT UA_StatusCode UA_THREADSAFE
 UA_Server_unfreezeWriterGroupConfiguration(UA_Server *server, const UA_NodeId writerGroup);
 
 UA_EXPORT UA_StatusCode UA_THREADSAFE
+UA_Server_enableWriterGroup(UA_Server *server, const UA_NodeId writerGroup);
```
Comment (jpfr) [inline include/open62541/server_pubsub.h:578]: Explain the difference between *enable* and *setOperational*.

We currently have both and redundancy leads to confusion.

**Review by jpfr — COMMENTED** [review]
A question wrt the API inline.
Thanks and keep up the good work.

File: include/open62541/server_pubsub.h (hunk 1 above)
Comment (gauravks) [reply include/open62541/server_pubsub.h:578]: The _enable_ API pushes the component state to Pre-Operational state if the parent of the Component is Operational and wait for trigger to Operational state. In the current implementation, there is no internal component for Reader/Writer Group and for DataSetReader which can push the state to Operational this needs to be done by the application explicitly using the _setOperational_.

in future the _setOperational_ can be removed in favour of a mechanism/internal pubsub component which pushes the state to Operational.

Comment (jpfr) [reply include/open62541/server_pubsub.h:578]: My question was:

We have both setOperational and enable as separate API methods.
What is the difference between them?

It seems to me that “enable” is redundant and will essentially do the same as setOperational (which also drops down to pre-operational if required).

File: src/pubsub/ua_pubsub_keystorage.c [hunk 2]
```diff
@@ -265,9 +265,11 @@ setPubSubGroupEncryptingKey(UA_Server *server, UA_NodeId PubSubGroupId, UA_UInt3
                             UA_ByteString signingKey, UA_ByteString encryptingKey,
                             UA_ByteString keyNonce) {
     UA_LOCK_ASSERT(&server->serviceMutex, 1);
+    UA_UNLOCK(&server->serviceMutex);
     UA_StatusCode retval =
-        setWriterGroupEncryptionKeys(server, PubSubGroupId, securityTokenId,
+        UA_Server_setWriterGroupEncryptionKeys(server, PubSubGroupId, securityTokenId,
```
Comment (jpfr) [inline src/pubsub/ua_pubsub_keystorage.c:270]: Continue to use `setWriterGroupEncryptionKeys`.
Then you don't have to unlock/lock.

File: src/pubsub/ua_pubsub_ns0.c [hunk 3]
```diff
@@ -1599,6 +1673,34 @@ addReaderGroupRepresentation(UA_Server *server, UA_ReaderGroup *readerGroup) {
...
+
+    UA_NodeId stateIdNode = 
+        findSingleChildNode(server, UA_QUALIFIEDNAME(0, "State"),
+                            UA_NODEID_NUMERIC(0, UA_NS0ID_HASCOMPONENT),
+                            statusIdNode);
+
+    if(UA_NodeId_isNull(&stateIdNode))
+        return UA_STATUSCODE_BADNOTFOUND;
+
+    UA_NodePropertyContext * stateContext = (UA_NodePropertyContext *)
```
Comment (jpfr) [inline src/pubsub/ua_pubsub_ns0.c:1693]: Always check if malloc returns NULL.

File: src/pubsub/ua_pubsub_ns0.c [hunk 4]
```diff
@@ -1810,6 +1939,14 @@ addDataSetWriterRepresentation(UA_Server *server, UA_DataSetWriter *dataSetWrite
     retVal |= addVariableValueSource(server, valueCallback,
                                      dataSetWriterIdNode, dataSetWriterIdContext);
 
+    UA_NodePropertyContext *dataSetWriterStateContext =
```
Comment (jpfr) [inline src/pubsub/ua_pubsub_ns0.c:1942]: Always check if malloc returns NULL.

File: src/pubsub/ua_pubsub_reader.c [hunk 5]
```diff
@@ -518,6 +517,34 @@ UA_Server_DataSetReader_getState(UA_Server *server, UA_NodeId dataSetReaderIdent
     return res;
 }
 
+UA_StatusCode
+UA_Server_enableDataSetReader(UA_Server *server, const UA_NodeId dataSetReaderIdentifier) {
+    UA_LOCK(&server->serviceMutex);
+    UA_StatusCode ret = UA_STATUSCODE_BADNOTFOUND;
+    UA_DataSetReader* dsr = UA_ReaderGroup_findDSRbyId(server, dataSetReaderIdentifier);
+
+    if(dsr)
```
Comment (jpfr) [inline src/pubsub/ua_pubsub_reader.c:526]: We prefer early-returns to deeply nested if-else.

File: src/pubsub/ua_pubsub_reader.c [hunk 6]
```diff
@@ -518,6 +517,34 @@ UA_Server_DataSetReader_getState(UA_Server *server, UA_NodeId dataSetReaderIdent
...
+    {
+        UA_NodeId rgId = dsr->linkedReaderGroup;
+        UA_ReaderGroup* readerGroup = UA_ReaderGroup_findRGbyId(server, rgId);
+        if(readerGroup->state == UA_PUBSUBSTATE_OPERATIONAL || readerGroup->state == UA_PUBSUBSTATE_PREOPERATIONAL) {
```
Comment (jpfr) [inline src/pubsub/ua_pubsub_reader.c:530]: Break overlong lines.
Here it is easy to do in the if statement.

File: src/pubsub/ua_pubsub_readergroup.c [hunk 7]
```diff
@@ -565,17 +591,32 @@ UA_Server_setReaderGroupOperational(UA_Server *server,
...
-        ret = UA_ReaderGroup_setPubSubState(server, rg, UA_PUBSUBSTATE_OPERATIONAL,
+
+UA_StatusCode
+UA_Server_enableReaderGroup(UA_Server *server, const UA_NodeId readerGroupId){
+    UA_LOCK(&server->serviceMutex);
+    UA_StatusCode ret = UA_STATUSCODE_BADNOTFOUND;
+    UA_ReaderGroup *rg = UA_ReaderGroup_findRGbyId(server, readerGroupId);
+    if(rg)
+    {
+        // TODO: Check connection state to decide the taget state
```
Comment (jpfr) [inline src/pubsub/ua_pubsub_readergroup.c:608]: Resolve the TODO.

File: src/pubsub/ua_pubsub_writer.c [hunk 8]
```diff
@@ -97,6 +97,155 @@ UA_DataSetWriterConfig_clear(UA_DataSetWriterConfig *pdsConfig) {
...
+                                    "Received unknown PubSub state!");
+    }
+
+    return ret;
+}
+
+static UA_StatusCode setPubSubState_paused(UA_Server *server,
+                                    UA_DataSetWriter *dataSetWriter) {
+    UA_StatusCode ret = UA_STATUSCODE_GOOD;
+    switch (dataSetWriter->state){
```
Comment (jpfr) [inline src/pubsub/ua_pubsub_writer.c:155]: This doesn't change that state at all!

**Review by jpfr — COMMENTED** [review]
Looks good.
Comments inline.

File: src/pubsub/ua_pubsub_keystorage.c (hunk 2 above)
Comment (gauravks) [reply src/pubsub/ua_pubsub_keystorage.c]: Thanks I have updated this to use `setWriterGroupEncryptionKeys`

File: src/pubsub/ua_pubsub_ns0.c [hunk 9]
```diff
@@ -1390,6 +1452,18 @@ addWriterGroupRepresentation(UA_Server *server, UA_WriterGroup *writerGroup) {
     writeAccessLevelAttribute(server, publishingIntervalNode,
                               UA_ACCESSLEVELMASK_READ ^ UA_ACCESSLEVELMASK_WRITE);
 
+    UA_NodePropertyContext * stateContext = (UA_NodePropertyContext *)
```
[the remark this answers is not in the document]
Comment (gauravks) [reply src/pubsub/ua_pubsub_ns0.c:1456]: Done

File: src/pubsub/ua_pubsub_ns0.c (hunk 3 above)
Comment (gauravks) [reply src/pubsub/ua_pubsub_ns0.c:1695]: Done

File: src/pubsub/ua_pubsub_ns0.c [hunk 10]
```diff
@@ -1799,6 +1902,32 @@ addDataSetWriterRepresentation(UA_Server *server, UA_DataSetWriter *dataSetWrite
...
+    }
+
+    UA_NodeId stateIdNode = 
+        findSingleChildNode(server, UA_QUALIFIEDNAME(0, "State"),
+                            UA_NODEID_NUMERIC(0, UA_NS0ID_HASCOMPONENT),
+                            statusIdNode);
+
+    if(UA_NodeId_isNull(&keyFrameNode)) {
+
+        // TODO Check why keyFrameNode is null
```
[the remark this answers is not in the document]
Comment (gauravks) [reply src/pubsub/ua_pubsub_ns0.c:1924]: The issue was present before. I included the check to test if the value is NULL. This could be a bug to be resolved later.
As discussed I would update the comment to make it a general issue

File: src/pubsub/ua_pubsub_ns0.c (hunk 4 above)
Comment (gauravks) [reply src/pubsub/ua_pubsub_ns0.c:1944]: Done

File: src/pubsub/ua_pubsub_reader.c (hunk 5 above)
Comment (gauravks) [reply src/pubsub/ua_pubsub_reader.c]: Thanks, I have updated it to remove one layer of nested if

File: src/pubsub/ua_pubsub_readergroup.c (hunk 7 above)
Comment (gauravks) [reply src/pubsub/ua_pubsub_readergroup.c:610]: This would be resolved in the pull request #5706

File: src/pubsub/ua_pubsub_writer.c (hunk 8 above)
Comment (gauravks) [reply src/pubsub/ua_pubsub_writer.c:155]: The paused state does not have any affect defined yet so there is no changes to be done except the state change variable updates

File: src/pubsub/ua_pubsub_writer.c [hunk 11]
```diff
@@ -97,6 +97,155 @@ UA_DataSetWriterConfig_clear(UA_DataSetWriterConfig *pdsConfig) {
...
+
+static UA_StatusCode setPubSubState_preoperational(UA_Server *server,
+                                            UA_DataSetWriter *dataSetWriter) {
+    UA_StatusCode ret = UA_STATUSCODE_GOOD;
+    switch (dataSetWriter->state){
+        case UA_PUBSUBSTATE_DISABLED:
+            dataSetWriter->state = UA_PUBSUBSTATE_PREOPERATIONAL;
+            break;
+        case UA_PUBSUBSTATE_PAUSED:
+            dataSetWriter->state = UA_PUBSUBSTATE_PREOPERATIONAL;
```
[the remark this answers is not in the document]
Comment (gauravks) [reply src/pubsub/ua_pubsub_writer.c]: The datasetwriter do no have any callback, this is part for the writergroup so there is nothing to be done here

File: src/pubsub/ua_pubsub_writergroup.c [hunk 12]
```diff
@@ -222,6 +222,14 @@ UA_WriterGroup_create(UA_Server *server, const UA_NodeId connection,
 
     if(writerGroupIdentifier)
         UA_NodeId_copy(&newWriterGroup->identifier, writerGroupIdentifier);
+
+    if (writerGroupConfig->enabled)
+    {
+        UA_UNLOCK(&server->serviceMutex);
```
[the remark this answers is not in the document]
Comment (gauravks) [reply src/pubsub/ua_pubsub_writergroup.c]: Okay will update

File: examples/pubsub/sks/server_pubsub_central_sks.c [hunk 13]
```diff
@@ -268,6 +268,7 @@ main(int argc, char **argv) {
         if(strcmp(argv[pos], "--port") == 0) {
             pos++;
             int inNum = atoi(argv[pos]);
+            printf("test\n");
```
Comment (jpfr) [inline examples/pubsub/sks/server_pubsub_central_sks.c:271]: Remove this debugging artifact.

File: src/pubsub/ua_pubsub_writer.c [hunk 14]
```diff
@@ -97,6 +99,107 @@ UA_DataSetWriterConfig_clear(UA_DataSetWriterConfig *pdsConfig) {
...
+                                                  UA_STATUSCODE_GOOD);
+        }
+    }
+    UA_UNLOCK(&server->serviceMutex);
+    return ret;
+}
+
+static UA_StatusCode
+startPubSubStateTimer(UA_Server *server, UA_DataSetWriter *dsw) {
+    UA_StatusCode ret = UA_STATUSCODE_BAD;
```
Comment (jpfr) [inline src/pubsub/ua_pubsub_writer.c:132]: Why do we need to wait 10ms to set the Writer operational?
Which change do we expect in the meantime and why can’t we set the new state immediately?
At the very least this should be documented in a comment.

File: include/open62541/server_pubsub.h [hunk 15]
```diff
@@ -552,7 +552,7 @@ UA_EXPORT UA_StatusCode UA_THREADSAFE
 UA_Server_unfreezeWriterGroupConfiguration(UA_Server *server, const UA_NodeId writerGroup);
 
 UA_EXPORT UA_StatusCode UA_THREADSAFE
-UA_Server_setWriterGroupOperational(UA_Server *server, const UA_NodeId writerGroup);
```
Comment (jpfr) [inline include/open62541/server_pubsub.h:555]: Don’t break existing code.
Please define the old UA_Server_setWriterGroupOperational as a macro alias for the new API.
You can add the UA_DEPRECATED attribute.
So developers get a compiler warning, but the code still works.
We can then remove the old API for the over ext stable release.

**Review by jpfr — COMMENTED** [review]
You have prepared this very nice table overview of the states.
This should be added to the top of ua_pubsub.h.

for the ReaderGroup, the semantics of preoperational is clear.
The rg is preoperational until the first message is received.

For the writergroup I am not so sure about the semantics.
Setting up a connection (e.g. mqtt) can take time.
So we set it preoperational until the connection has fully opened.
Setting it operational is done within the network-callback where the open connection is signaled. Do you share this understanding?

otherwise this is coming together nicely. If you add the table, we can merge in a matter of hours.

File: include/open62541/server_pubsub.h (hunk 15 above)
Comment (gauravks) [reply include/open62541/server_pubsub.h:555]: Okay will do that

File: src/pubsub/ua_pubsub_writer.c (hunk 14 above)
Comment (gauravks) [reply src/pubsub/ua_pubsub_writer.c:132]: Currently there is no change expected and the writer can go immediately to operational state. I added this as a future considerations when there is some checks to be performed.

I would add a comment to make this clear

File: include/open62541/server_pubsub.h [hunk 16]
```diff
@@ -845,7 +853,7 @@ UA_EXPORT UA_StatusCode UA_THREADSAFE
 UA_Server_unfreezeReaderGroupConfiguration(UA_Server *server, const UA_NodeId readerGroupId);
 
 UA_EXPORT UA_StatusCode UA_THREADSAFE
-UA_Server_setReaderGroupOperational(UA_Server *server, const UA_NodeId readerGroupId);
```
[the remark this answers is not in the document]
Comment (gauravks) [reply include/open62541/server_pubsub.h:848]: I will add this

**Comment by gauravks** [comment]
> For the writergroup I am not so sure about the semantics.
> Setting up a connection (e.g. mqtt) can take time.
> So we set it preoperational until the connection has fully opened.
> Setting it operational is done within the network-callback where the open connection is signaled. Do you share this understanding?

Yes my understanding is the same now with the eventloop. An additional consideration which I have included is regarding the setting up of SecurityKeys. For an SKS based system, The security keys may be set later, till then the WriterGroup stays in preoperational.

**Comment by gauravks** [comment]
> You have prepared this very nice table overview of the states.
> This should be added to the top of ua_pubsub.h

Thanks. I will include this, I was not sure if this goes as Markdown or incode. Now it is clear

**Comment by gauravks** [comment]
I have updated the pull request with the documentation and integrated the comments

File: examples/pubsub_realtime/pubsub_TSN_publisher_multiple_thread.c [hunk 17]
```diff
@@ -1414,13 +1416,12 @@ int main(int argc, char **argv) {
     addWriterGroup(server);
     addDataSetWriter(server);
     UA_Server_freezeWriterGroupConfiguration(server, writerGroupIdent);
-    UA_Server_setWriterGroupOperational(server, writerGroupIdent);
```
Comment (andreasebner) [inline examples/pubsub_realtime/pubsub_TSN_publisher_multiple_thread.c:1417]: Why is here no enable logic needed? (whole file)

File: src/pubsub/ua_pubsub_ns0.c [hunk 18]
```diff
@@ -1792,6 +1898,31 @@ addDataSetWriterRepresentation(UA_Server *server, UA_DataSetWriter *dataSetWrite
...
+    }
+
+    UA_NodeId stateIdNode = 
+        findSingleChildNode(server, UA_QUALIFIEDNAME(0, "State"),
+                            UA_NODEID_NUMERIC(0, UA_NS0ID_HASCOMPONENT),
+                            statusIdNode);
+
+    if(UA_NodeId_isNull(&keyFrameNode)) {
+
+        // TODO: The keyFrameNode is NULL here, does not depend on the pubsub changes
```
Comment (andreasebner) [inline src/pubsub/ua_pubsub_ns0.c:1917]: If we do nothing here, a simple comment is fine.

File: src/pubsub/ua_pubsub_writer.c [hunk 19]
```diff
@@ -97,6 +99,113 @@ UA_DataSetWriterConfig_clear(UA_DataSetWriterConfig *pdsConfig) {
...
+UA_Server_enableDataSetWriter(UA_Server *server, const UA_NodeId dataSetWriterIdent) {
+    UA_LOCK(&server->serviceMutex);
+    UA_StatusCode ret = UA_STATUSCODE_BADNOTFOUND;
+    UA_DataSetWriter *dataSetWriter = UA_DataSetWriter_findDSWbyId(server, dataSetWriterIdent);
+
+    if(dataSetWriter)
+    {
+        UA_NodeId wgId = dataSetWriter->linkedWriterGroup;
+        UA_WriterGroup* writerGroup = UA_WriterGroup_findWGbyId(server, wgId);
+        if(writerGroup->state == UA_PUBSUBSTATE_OPERATIONAL || writerGroup->state == UA_PUBSUBSTATE_PREOPERATIONAL) {
```
Comment (andreasebner) [inline src/pubsub/ua_pubsub_writer.c:112]: Always check if "find" is successful.

**Review by andreasebner — COMMENTED** [review]
Thanks’ for the great work. I made my final review and added some minor comments. From my point of view the PR is ready to merge afterwards. 

The dedicated unit-test file for the states has not grown a lot. Maybe this is the reason for the failing code coverage CI. You need to add more tests (CI allows a maximum decrease of -0.1%). 

@jpfr after these points are solved, we should merge the PR.

File: examples/pubsub_realtime/pubsub_TSN_publisher_multiple_thread.c (hunk 17 above)
Comment (gauravks) [reply examples/pubsub_realtime/pubsub_TSN_publisher_multiple_thread.c:1417]: I have moved enableWriterGroup call to addWriterGroup() function

File: src/pubsub/ua_pubsub_ns0.c (hunk 18 above)
Comment (gauravks) [reply src/pubsub/ua_pubsub_ns0.c:1917]: Okay

File: src/pubsub/ua_pubsub_writer.c (hunk 19 above)
Comment (gauravks) [reply src/pubsub/ua_pubsub_writer.c:112]: Thanks I will include the check
